### PR TITLE
Establish CoreDataStack mock usage pattern - Part I

### PR DESCRIPTION
**Pull request by crazytonyli**
This PR is part of paaHJt-3ky-p2.

## Changes
The ultimate goal is making `ContextManagerMock` the only mock implementation of `CoreDataStack`. This PR is first step towards that goal.

A couple changes involved:
* Removes `TextContextManager.sharedInstance`. I don't think this singleton is particularly useful. By removing it, `TestContextManager` is now created in each individual test case, this will simplify the clean up work down the road.
* Moves overriding `ContextManager.sharedInstance` logic from `ContextManagerMock` to `TestContextManager`. This will help us understand which test case relies on the `ContextManager` singleton, when later all `TestContextManager` usage is replaced by `ContextManagerMock`.

## Test Instructions
Make sure all unit tests pass.

## Regression Notes
N/A. All changes in this PR are in unit test, they don't have any impact on the app.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/WordPressTest/TestContextManager.m
Comment (mokagio) [inline WordPress/WordPressTest/TestContextManager.m:27]: The reason we're calling this is... so that we ensure the internal instance (`_instance`) is set then overridden by us. Otherwise, any call to `[ContextManager internalSharedInstance]` by the code under test would run the `dispatch_once` and return a fresh `_instance`, bypassing our override. Is that correct?

https://github.com/wordpress-mobile/WordPress-iOS/blob/5f4e7696e2745ed4d14e09670ab2f84670cbcedc/WordPress/Classes/Utility/ContextManager.m#L43-L51

File: WordPress/WordPressTest/Classes/Stores/ActivityStoreTests.swift
Comment (mokagio) [inline WordPress/WordPressTest/Classes/Stores/ActivityStoreTests.swift:25]: These look odd, but will go away in your next step which will add dedicated `setUp...` and `tearDown` method to the test double, right?

I say they look odd because, with my reader hat on, it's not straightforward to understand why one would want to override the shared instance with `nil`.

File: WordPress/WordPressTest/ReaderPostTest.m
Comment (mokagio) [inline WordPress/WordPressTest/ReaderPostTest.m]: Nitpick: [We plan to stop calling `super` in the life-cycle overrides.](https://github.com/wordpress-mobile/WordPress-iOS/issues/17569):

> There's nothing wrong in calling `super` in setup or teardown methods, but it's unnecessary because `XCTestCase` doesn't implement any functionality in there, making calling `super` redundant.
>
> Note that I couldn't find any _explicit_ proof in the docs to back my assertion. But, I'll point you towards [the life-cycle documentation page](https://developer.apple.com/documentation/xctest/xctestcase/understanding_setup_and_teardown_for_test_methods) to see that there is no mention of calling the superclass. 

```suggestion
    self.coreDataStack = [[TestContextManager alloc] init];
}

- (void)tearDown
{
    self.coreDataStack = nil;
```

We could keep them, just for consistency with the current state of things, although I'd argue against it because it only adds extra work to our future selves. 

If we decided to keep them, what do you think of moving the `[super tearDown]` call to the end of the method? That's how I used to see it done elsewhere in the past: first cleanup our code, then let `super` do it's work.

```suggestion
    [super setUp];
    self.coreDataStack = [[TestContextManager alloc] init];
}

- (void)tearDown
{
    self.coreDataStack = nil;
    [super tearDown];
```

**Review by mokagio — APPROVED** [review]
Only one comment about calling `setUp` and `tearDown` methods that are possibly unnecessary.

I'm going to approve this to unblock you. So, if I'm not around to click the merge button when that's been addressed, you can quickly get any maintainer to merge by referencing this review 👍

File: WordPress/WordPressTest/TestContextManager.m
Comment (crazytonyli) [reply WordPress/WordPressTest/TestContextManager.m:27]: The short answer is, I kept the [original code](https://github.com/wordpress-mobile/WordPress-iOS/pull/18512/files#diff-7b102973d3b2dcb347034552ed9c073ba74e2a870795fc06ad671ec905fd6ce9) in `ContextManagerMock`.

As for the longer version... It's a bit complicated, I'll try to explain...

------

To understand why this is needed, we'll need to go back to the dispatch_once calls in [`ContextManager.mainContext`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L75) and [`writerContext`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L63).

The issue they have caused is, only _the very first instance_ that calls `mainContext` or `writerContext` gets a valid value, all other instances get nil. Bear in mind, the "instance" here also includes instances of `ContextManager` subclass `ContextManagerMock`.

Let's discuss a hypothetical situation first. Say there are two such instances created at runtime: fakeManagerA and realManagerB. And following calls are made, in their listed order:
1. `fakeManagerA.writerContext`
    As a result, fakeManagerA gets a valid `writerContext` value.
1. `realManagerB.writerContext`.
    As a result, realManagerB gets a nil `writerContext`, since fakeManagerA called `writerContext` first.
1. `realManagerB.mainContext`.
    BOOM. Crash. Due to exception "Parent NSManagedObjectContext must not be nil".
    Why? Because `mainContext` [uses `writerContext` as its parent context](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L77), which is nil as mentioned above.

Of course, this isn't a real issue in the app, for two facts:
1. There is only one single instance of `ContextManager`.
1. `writerContext` is a private API, no way the first call to it can be made by anyone else other than `mainContext`.

Unfortunately both are not facts in unit test.
1. Test cases can create `TestContextManager` whenever they want, which means there are [many instances of `ContextManagerMock`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/WordPressTest/TestContextManager.m#L16), subclass of `ContextManager`.
2. `ContextManagerMock` [overrides `mainContext`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/WordPressTest/ContextManagerMock.m#L67-L77), which means _`ContextManager.mainContext`_ method is _not called_ when the instance is a `ContextManagerMock`, so it's very much possible that `ContextManager.writerContext` is called _without calling_ `ContextManager.mainContext`, like by [this notification handler](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L258-L265).

Now, let's go back to the hypothetical situation discussed before. Consider fakeManagerA as an instance of `ContextManagerMock`, realManagerB as the `ContextManager` singleton. And following calls are made, in their listed order:
1. `fakeManagerA.mainContext` is called.
    As mentioned above, this method returns the context created by `ContextManagerMock.mainContext` override, not `ContextManager.mainContext`.
1. `fakeManagerA.mainContext.save()` is called.
    As a side effect of this, [the notification hander](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L258-L265) is now called by CoreData framework.
1. `fakeManagerA.writerContext` is called by above notification handler.
    As a result, fakeManagerA gets a valid `writerContext` value.
1. (Looks familiar? Yes, you have seen this "~hypothetical~ situation" before)
1. `realManagerB.writerContext` is called.
    As a result, realManagerB gets a nil value, since fakeManagerA calls writerContext first.
1. `realManagerB.mainContext` is called. BOOM. Unit test crashes, for reason discussed before.

A workaround for this is, making sure realManagerB, the `ContextManager` singleton, calls `mainContext` and `writerContext` first. Which is why you see this statement `[ContextManager internalSharedInstance]` here, to force creating `ContextManager` singleton[^1].

Apologise this turned into an essay... But I appreciate you reading this far, what an adventure!

[^1]: As I mentioned in the P2 post(paaHJt-3ky-p2), both `mainContext` and `writerContext` are created during initialisation, which means their `dispatch_once` calls are redundant.

File: WordPress/WordPressTest/Classes/Stores/ActivityStoreTests.swift
Comment (crazytonyli) [reply WordPress/WordPressTest/Classes/Stores/ActivityStoreTests.swift:25]: Yes, you are absolutely correct. This is the pattern currently used in test cases, which I hope to improve as result of this and its follow-up PRs.

File: WordPress/WordPressTest/ReaderPostTest.m
Comment (crazytonyli) [reply WordPress/WordPressTest/ReaderPostTest.m]: I don't really have a strong opinion on this. For me, I think it's a habit that I developed over the years - always call super at beginning of the function, unless the API doc explicitly says don't call it or expected placement of super call. I can definitely remove those super calls added in this PR due to my personal habit 😄 

If this is something we'd like to do going forward, maybe it's worth considering adding a swiftlint rule for it, like:
```yaml
prohibited_super_call:
  included:
    - setUp()
    - tearDown()
```

File: WordPress/WordPressTest/TestContextManager.m
Comment (mokagio) [reply WordPress/WordPressTest/TestContextManager.m:27]: > Apologise this turned into an essay... But I appreciate you reading this far, what an adventure!

It was a fun read. Thank you 😄

**Pull request merged**
